src/hotspot/share/gc/z/zBitMap.inline.hpp

```diff
@@ -47,26 +47,29 @@ inline bool ZBitMap::par_set_bit_pair_strong(idx_t bit, bool& inc_live) {
   verify_index(bit);
   volatile bm_word_t* const addr = word_addr(bit);
   const bm_word_t pair_mask = bit_mask_pair(bit);
-  bm_word_t old_val = *addr;
+  bm_word_t cur_val = *addr;
 
-  do {
-    const bm_word_t new_val = old_val | pair_mask;
-    if (new_val == old_val) {
-      // Someone else beat us to it
-      inc_live = false;
-      return false;
-    }
-    const bm_word_t cur_val = Atomic::cmpxchg(addr, old_val, new_val);
-    if (cur_val == old_val) {
-      // Success
-      const bm_word_t marked_mask = bit_mask(bit);
-      inc_live = !(old_val & marked_mask);
-      return true;
-    }
+  // Fast path
+  bm_word_t new_val = cur_val | pair_mask;
+  if (new_val == cur_val) {
+    // Someone else beat us to it
+    inc_live = false;
+    return false;
+  }
+
+  // Slow path
+  cur_val = Atomic::fetch_and_or(addr, pair_mask);
+  new_val = cur_val | pair_mask;
+  if (new_val == cur_val) {
+    // Someone else beat us to it
+    inc_live = false;
+    return false;
+  }
```

Comment:
The fast/slow paths are almost identical. I believe some numbers would be nice to justify the presence of the fast path.

Reply:
I think so. Atomic operations are more expensive than normal memory loads. This is why should split to fast and slow path, although they are almost Identical. The micro benchmark 2 shows it.

Reply:
Thank you for the benchmark results; they are convincing, IMO.